File: sidecar/src/backup.ts
import { cp, copyFile, lstat, readdir, rm, stat } from "node:fs/promises";
import { basename, dirname, isAbsolute, join, relative, resolve } from "node:path";
import { ConvertZZError } from "./errors.js";

export interface BackupRoot {
  path: string;
  kind: "file" | "directory";
}

/** 來源路徑對應的使用者備份路徑：`path.bak`。 */
export function userBackupPath(sourcePath: string): string {
  return `${sourcePath}.bak`;
}

/**
 * 依計畫輸入路徑決定備份單位。
 * 選取資料夾時整份備份為 `資料夾.bak`；選取檔案或萬用字元展開的檔案則逐檔 `.bak`。
 */
export async function resolveBackupRoots(paths: string[]): Promise<BackupRoot[]> {
  const roots: BackupRoot[] = [];
  for (const path of paths) {
    if (/[*?]/u.test(path)) {
      for (const match of await expandWildcardFiles(path)) {
        roots.push({ path: match, kind: "file" });
      }
      continue;
    }
    const absolute = resolve(path);
    try {
      const info = await lstat(absolute);
      if (info.isSymbolicLink()) continue;
      if (info.isDirectory()) roots.push({ path: absolute, kind: "directory" });
      else if (info.isFile()) roots.push({ path: absolute, kind: "file" });
    } catch {
      // 計畫階段可能含尚未驗證的路徑，略過無法備份的項目。
    }
  }
  return pruneNestedBackupRoots(roots);
}

/** 只保留最外層資料夾與不在其內的檔案，避免同一樹重複備份。 */
export function pruneNestedBackupRoots(roots: BackupRoot[]): BackupRoot[] {
  const unique = new Map<string, BackupRoot>();
  for (const root of roots) unique.set(root.path, root);
  const list = Array.from(unique.values());
  const directories = list
    .filter((root) => root.kind === "directory")
    .map((root) => root.path)
    .sort((left, right) => left.length - right.length);

  return list.filter((root) => {
    const coveredByOuterDirectory = directories.some(
      (directory) => directory !== root.path && pathIsInside(root.path, directory),
    );
    return !coveredByOuterDirectory;
  });
}

/**
 * 對會被修改的來源建立 `.bak`。
 * 資料夾根：整份複製為 `folder.bak`；檔案根：`file.ext.bak`。
 */
export async function createUserBackups(
  roots: BackupRoot[],
  affectedPaths: string[],
): Promise<string[]> {
  if (!roots.length || !affectedPaths.length) return [];
  const affected = affectedPaths.map((path) => resolve(path));
  const created: string[] = [];

  for (const root of roots) {
    const covers = affected.some((path) =>
      root.kind === "directory" ? pathIsInside(path, root.path) : path === root.path,
    );
    if (!covers) continue;
    created.push(await createUserBackup(root.path));
  }
  return created;
}

export async function createUserBackup(sourcePath: string): Promise<string> {
  const absolute = resolve(sourcePath);
  const target = userBackupPath(absolute);
  try {
    const info = await lstat(absolute);
    if (info.isSymbolicLink()) {
      throw new ConvertZZError("BACKUP_SYMLINK", "不備份符號連結來源。", { path: absolute });
    }
    await rm(target, { recursive: true, force: true });
    if (info.isDirectory()) {
      await cp(absolute, target, {
        recursive: true,
        force: true,
        errorOnExist: false,
        verbatimSymlinks: true,
      });
    } else if (info.isFile()) {
      await copyFile(absolute, target);
    } else {
      throw new ConvertZZError("BACKUP_UNSUPPORTED", "不支援的備份來源類型。", { path: absolute });
    }
    return target;
  } catch (error) {
    if (error instanceof ConvertZZError) throw error;
    throw new ConvertZZError(
      "BACKUP_FAILED",
      `建立備份失敗：${error instanceof Error ? error.message : String(error)}`,
      { path: absolute, backupPath: target },
    );
  }
}

export function pathIsInside(path: string, directory: string): boolean {
  const relativePath = relative(resolve(directory), resolve(path));
  return relativePath === "" || (!relativePath.startsWith("..") && !isAbsolute(relativePath));
}

async function expandWildcardFiles(pattern: string): Promise<string[]> {
  const absolute = resolve(pattern);
  const directory = dirname(absolute);
  const matcher = wildcardMatcher(basename(absolute));
  try {
    await stat(directory);
  } catch {
    return [];
  }
  const entries = await readdir(directory, { withFileTypes: true });
  return entries
    .filter((entry) => entry.isFile() && matcher.test(entry.name))
    .map((entry) => join(directory, entry.name))
    .sort();
}

function wildcardMatcher(pattern: string): RegExp {
  const source = Array.from(pattern, (character) => {
    if (character === "*") return "(.*?)";
    if (character === "?") return "(.)";
    return character.replace(/[\\^$.*+?()[\]{}|]/gu, "\\$&");
  }).join("");
  return new RegExp(`^${source}$`, process.platform === "win32" ? "iu" : "u");
}

File: src-tauri/src/core/backup.rs
use super::error::CoreError;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BackupRoot {
    pub path: PathBuf,
    pub kind: BackupKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BackupKind {
    File,
    Directory,
}

pub fn user_backup_path(source: &Path) -> PathBuf {
    let mut path = source.as_os_str().to_os_string();
    path.push(".bak");
    PathBuf::from(path)
}

pub fn resolve_backup_roots(paths: &[String]) -> Result<Vec<BackupRoot>, CoreError> {
    let mut roots = Vec::new();
    for path in paths {
        if path.contains(['*', '?']) {
            roots.extend(
                expand_wildcard_files(path)?
                    .into_iter()
                    .map(|item| BackupRoot {
                        path: item,
                        kind: BackupKind::File,
                    }),
            );
            continue;
        }
        let absolute = resolve_path(path);
        let Ok(metadata) = fs::symlink_metadata(&absolute) else {
            continue;
        };
        if metadata.file_type().is_symlink() {
            continue;
        }
        if metadata.is_dir() {
            roots.push(BackupRoot {
                path: absolute,
                kind: BackupKind::Directory,
            });
        } else if metadata.is_file() {
            roots.push(BackupRoot {
                path: absolute,
                kind: BackupKind::File,
            });
        }
    }
    Ok(prune_nested_backup_roots(roots))
}

pub fn prune_nested_backup_roots(roots: Vec<BackupRoot>) -> Vec<BackupRoot> {
    let mut unique = Vec::new();
    for root in roots {
        if !unique
            .iter()
            .any(|item: &BackupRoot| item.path == root.path)
        {
            unique.push(root);
        }
    }
    let list = unique;
    let mut directories = list
        .iter()
        .filter(|root| root.kind == BackupKind::Directory)
        .map(|root| root.path.clone())
        .collect::<Vec<_>>();
    directories.sort_by_key(|path| path.as_os_str().len());
    list.into_iter()
        .filter(|root| {
            !directories
                .iter()
                .any(|directory| directory != &root.path && path_is_inside(&root.path, directory))
        })
        .collect()
}

pub fn create_user_backups(
    roots: &[BackupRoot],
    affected_paths: &[PathBuf],
) -> Result<Vec<PathBuf>, CoreError> {
    if roots.is_empty() || affected_paths.is_empty() {
        return Ok(Vec::new());
    }
    let affected = affected_paths
        .iter()
        .map(|path| resolve_path(&path.to_string_lossy()))
        .collect::<Vec<_>>();
    let mut created = Vec::new();
    for root in roots {
        let covers = affected.iter().any(|path| match root.kind {
            BackupKind::Directory => path_is_inside(path, &root.path),
            BackupKind::File => path == &root.path,
        });
        if covers {
            created.push(create_user_backup(&root.path)?);
        }
    }
    Ok(created)
}

pub fn create_user_backup(source_path: &Path) -> Result<PathBuf, CoreError> {
    let absolute = resolve_path(&source_path.to_string_lossy());
    let target = user_backup_path(&absolute);
    let metadata = fs::symlink_metadata(&absolute)?;
    if metadata.file_type().is_symlink() {
        return Err(CoreError::with_details(
            "BACKUP_SYMLINK",
            "不備份符號連結來源。",
            serde_json::json!({ "path": absolute }),
        ));
    }
    let _ = fs::remove_dir_all(&target);
    let _ = fs::remove_file(&target);
    if metadata.is_dir() {
        copy_dir(&absolute, &target)?;
    } else if metadata.is_file() {
        fs::copy(&absolute, &target)?;
    } else {
        return Err(CoreError::with_details(
            "BACKUP_UNSUPPORTED",
            "不支援的備份來源類型。",
            serde_json::json!({ "path": absolute }),
        ));
    }
    Ok(target)
}

pub fn path_is_inside(path: &Path, directory: &Path) -> bool {
    let path = resolve_path(&path.to_string_lossy());
    let directory = resolve_path(&directory.to_string_lossy());
    path == directory || path.starts_with(&directory)
}

fn copy_dir(source: &Path, destination: &Path) -> Result<(), CoreError> {
    fs::create_dir_all(destination)?;
    for entry in fs::read_dir(source)? {
        let entry = entry?;
        let from = entry.path();
        let to = destination.join(entry.file_name());
        let metadata = fs::symlink_metadata(&from)?;
        if metadata.file_type().is_symlink() {
            #[cfg(unix)]
            {
                let target = fs::read_link(&from)?;
                let _ = fs::remove_file(&to);
                std::os::unix::fs::symlink(target, &to)?;
            }
        } else if metadata.is_dir() {
            copy_dir(&from, &to)?;
        } else if metadata.is_file() {
            fs::copy(&from, &to)?;
        }
    }
    Ok(())
}

fn expand_wildcard_files(pattern: &str) -> Result<Vec<PathBuf>, CoreError> {
    let absolute = resolve_path(pattern);
    let directory = absolute.parent().unwrap_or(Path::new("."));
    let Some(name) = absolute.file_name().and_then(|name| name.to_str()) else {
        return Ok(Vec::new());
    };
    if fs::metadata(directory).is_err() {
        return Ok(Vec::new());
    }
    let matcher = wildcard_matcher(name);
    let mut matches = Vec::new();
    for entry in fs::read_dir(directory)? {
        let entry = entry?;
        if entry.file_type()?.is_file() {
            if let Some(file_name) = entry.file_name().to_str() {
                if matcher.is_match(file_name) {
                    matches.push(entry.path());
                }
            }
        }
    }
    matches.sort();
    Ok(matches)
}

pub fn wildcard_matcher(pattern: &str) -> regex::Regex {
    let mut source = String::from("^");
    for character in pattern.chars() {
        match character {
            '*' => source.push_str("(.*?)"),
            '?' => source.push_str("(.)"),
            other => source.push_str(&regex::escape(&other.to_string())),
        }
    }
    source.push('$');
    let flags = if cfg!(windows) { "(?iu)" } else { "(?u)" };
    regex::Regex::new(&format!("{flags}{source}")).expect("wildcard")
}

pub fn resolve_path(path: &str) -> PathBuf {
    let candidate = PathBuf::from(path);
    if candidate.is_absolute() {
        candidate
    } else {
        std::env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("."))
            .join(candidate)
    }
}

#[cfg(test)]
mod tests;
